internal/host/plugin/repository_host_catalog.go

```diff
@@ -25,6 +26,38 @@ import (
 	"google.golang.org/protobuf/types/known/wrapperspb"
 )
 
+// normalizeCatalogAttributes allows a plugin to normalize attributes before
+// they are saved
+func normalizeCatalogAttributes(ctx context.Context, plgClient plgpb.HostPluginServiceClient, plgHc *pb.HostCatalog) error {
+	const op = "plugin.(Repository).normalizeCatalogAttributes"
+	switch {
+	case util.IsNil(plgClient):
+		return errors.New(ctx, errors.InvalidParameter, op, "plugin client is nil")
+	case plgHc == nil:
+		return errors.New(ctx, errors.InvalidParameter, op, "host catalog is nil")
+	case plgHc.GetAttributes() == nil:
+		return nil
+	}
+
+	ret, err := plgClient.NormalizeCatalogData(ctx, &plgpb.NormalizeCatalogDataRequest{
+		Attributes: plgHc.GetAttributes(),
+	})
+	switch {
+	case err == nil:
+		if ret.Attributes != nil {
+			plgHc.Attrs = &pb.HostCatalog_Attributes{
+				Attributes: ret.Attributes,
+			}
+		}
+	case status.Code(err) == codes.Unimplemented:
+		// Do nothing
+	default:
+		return errors.Wrap(ctx, err, op, errors.WithMsg("error asking plugin to normalize catalog data"))
+	}
+
+	return nil
+}
```

Comment:
Nit: perhaps a little nicer than mutating the `plghc` input would be to take a `attributes` input and return a new `attributes`? In the "do nothing" case we could just return the input.

Reply:
I'm not sure if that's nicer or not? I guess the way I see it is in the create and update functions we're constantly mutating the object we're building up before saving. I don't really have a preference though.

Reply:
Fair enough, consistency is important 👍🏻